_posts/2020-12-10-webpack-vs-sprockets.markdown

```diff
@@ -0,0 +1,313 @@
+
+When using Sprockets, all JavaScript is concatenated in one plain file and everything runs in the global scope. So you can call `initMap` from anywhere by default. This is really handy... BUT! there's is a problem: it bloats the global scope and you can have different modules using the same function causing name collision.
+
+When using Webpack, each script is isolated so nothing changes the global scope by default and only exports the things you tell it to export (it could be a class, a function, an object, etc..., we can export multiple things too). This solves the problem of global scope contamination, but you won't be able to access the `initMap` function from anywhere since it's not exposed globally.
+
+To do that, you have to be explicit when defining the function in your module:
+
+```js
+global.initMap = function(mapId) {
+  // initializes a map plugin
+}
+
+// or
+
+const initMap = (mapId) => {
+  // initializes a map plugin
+}
+global.initMap = initMap
+
+```
+
+## jQuery
+
+Many projects depend on **jQuery** and needs the `$` function available everywhere. The easiest way to handle this is to tell Webpack to expose the `$` and the `jQuery` functions globally. To do that, you need to add a plugin setting:
+
+```js
+// config/webpack/environment.js
+const { environment } = require('@rails/webpacker')
+
+const webpack = require('webpack')
+environment.plugins.prepend(
+  'Provide',
+  new webpack.ProvidePlugin({
+    $: 'jquery/src/jquery',
+    jQuery: 'jquery/src/jquery'
+  })
+)
+
+module.exports = environment
+```
+
+## Livereload
+
+Webpacker comes with a handy bin file you can run to make your web auto-reload when you change any asset, so you don't have to refresh manually after each change you do. You can run `rails s` in one terminal and `bin/webpack-dev-server` in another terminal, enter `localhost:3000` and you can change any asset and your page will be updated with no need of `F5`. This is really handy when you need to do heavy assets work.
+
+## Non-JavaScript Assets With Webpacker
+
+### CSS
+
+When using Webpacker, while you can have CSS entry points at `app/javascript/packs`, it can create some problems and sometimes you will see something like this inside a JavaScript file instead:
+
+```js
+import 'my_file.css';
+```
+
+or even more strange-looking:
+
+```js
+import 'my_image.png';
+```
+
+What's happening here is that Webpacker can extract different type of assets referenced by your JavaScript packs and emit those as separated files! If you have an `application.js` file that imports some CSS, it will emit an `application.css` file with the content of that CSS. You have to be sure you have the right configuration:
+
+```js
+// config/webpacker.yml
+
+production:
+  extract_css: true
+```
+
+You can have that as `false` during development or testing, but it's needed for production. I'd recommend you at least try with this as `true` during development to verify it will create the right files for production.
```

Comment:
```suggestion
You can set that as `false` during development or testing, but it's needed for production. I'd recommend you at least try it as `true` during development to verify that it will create the right files for production.
```